File: code/nerdleGameSolver.cpp
#include <iostream>
#include <fstream>
#include <cmath>
#include <string>
#include <map>
#include <vector>
#include <utility>
#include <algorithm>

using namespace std;

string possibleSolutions8[17723];
map<string, int> PxAuxiliarMatchingTable; //key=Pattern+Expresion, value: numberOfMatchingExpressions

char simbolosPatronSimilitud[3] = {'0', '1', '2'};
string patterns[6561];
int numberOfTiles = 8;
int indexOfPattern = 0;

void createPatternsRecursionFunction(string prevString, char simbol)
{
    string newString = prevString += simbol;
    if (newString.length() < ::numberOfTiles)
    {
        for (int i = 0; i < 3; i++)
        {
            createPatternsRecursionFunction(newString, simbolosPatronSimilitud[i]);
        }
    }
    else
    {
        ::patterns[::indexOfPattern] = newString;
        ::indexOfPattern += 1;
    }
}

string generatePattern(int idxE1, int idxE2){
    string pattern="";
    for (int i = 0; i < 8; i++)
    {
        if(possibleSolutions8[idxE1]==possibleSolutions8[idxE2]) pattern+="2";
        else if(possibleSolutions8[idxE1].find(possibleSolutions8[idxE2][i]) != std::string::npos) pattern+="1";
        else pattern+="0";
    }
    return pattern;
}

void preComputePxAuxiliarTable() {
    for(int idxExpresionFija = 0; idxExpresionFija<17723; idxExpresionFija++){
        if(idxExpresionFija%500==0)cout<<idxExpresionFija<<"/17723"<<endl;
        for(int idxExpresionMovil = 0; idxExpresionMovil<17723; idxExpresionMovil++){
            string pattern = generatePattern(idxExpresionFija, idxExpresionMovil);
            PxAuxiliarMatchingTable[pattern+possibleSolutions8[idxExpresionFija]]++;
        }
    }
}

float calculatePx(string pattern, string currentExpression)
{
    if(PxAuxiliarMatchingTable.find(pattern+currentExpression)!=PxAuxiliarMatchingTable.end()) return (float)PxAuxiliarMatchingTable[pattern+currentExpression] / 17723;
    else return 0;
}

float calculateInformationVariables(string expression, string possibleSolutions[17723], string patterns[6561])
{
    /*
    E[i] = sumatoria(p(x)*log2(1/p(x)))
    Donde:
    x = patron de similitud
    p(x) = (N de eq posibles con el patron x dado) / (N total de eq)
     */

    float EinfoOfExpression = 0.0f;

    for (int i = 0; i < 6561; i++)
    {
        //cout << patterns[i] << "  Einfo: " << EinfoOfExpression << endl;
        float Px = calculatePx(patterns[i], expression);

        if (Px != 0.0f)
        {
            EinfoOfExpression += Px * log2(1 / Px);
        }
    }

    return EinfoOfExpression;
}

bool customSortFunc(string info1[2], string info2[2]){
    return true;
}

int main()
{
    // Reading possible solutions
    string solutionsText;

    ifstream MyReadFile("validSolutions8.txt");

    int indexOfSolutionReading = 0;
    while (getline(MyReadFile, solutionsText))
    {
        possibleSolutions8[indexOfSolutionReading] = solutionsText;
        indexOfSolutionReading++;
    }
    MyReadFile.close();

    // Generating array of possible patterns
    for (int i = 0; i < 3; i++)
    {
        createPatternsRecursionFunction("", simbolosPatronSimilitud[i]);
    }

    cout<<"STARTED precomputing PxAuxiliarTable"<<endl; //LOGGING
    preComputePxAuxiliarTable();
    cout<<"FINISHED precomputing PxAuxiliarTable"<<endl;

    cout<<"STARTED computing Einfo of expressions"<<endl;
    // Calculating information given every equation
    vector<pair<float, int>> informationList; 

    for (int i = 0; i < 17723; i++)
    {
        if(i%500==0) cout<<i<<"/17723"<<endl; //LOGGING

        string expression = possibleSolutions8[i];
        float EInfoOfExpression = calculateInformationVariables(expression, possibleSolutions8, patterns);

        informationList.push_back({EInfoOfExpression, i});
    }

    sort(informationList.rbegin(), informationList.rend());

    //saving data in .txt
    ofstream OutputFile("GuessesEinfo8.txt");
    for (int i = 0; i < 17723; i++)
    {
        OutputFile<<possibleSolutions8[informationList[i].second]<<": "<<informationList[i].first<<endl;
    }
    OutputFile.close();
    
    return 0;
}

File: code/nerdleGameSolver.py
""" 
Before running there must be a txt file containing all the valid Inputs. This file is created by 
running "possibleInputGenerator.py" and "validInputsGenerator.py".
"""

import math
import itertools
import os


def createInfoTxt(infoList):
    if os.path.exists("infoList8.txt"):
        os.remove("infoList8.txt")

    with open("infoList8.txt", "a") as f:
        for expression in infoList:
            f.write("{} {}\n".format(expression[0], expression[1]))

        pass
    return


def checkPattern(pattern, currentExpression, expression):

    for index, similitud in enumerate(pattern):
        if similitud == "0" and (currentExpression[index] in expression):
            return False
        elif similitud == "1" and (currentExpression[index] == expression[index] or currentExpression[index] not in expression):
            return False
        elif similitud == "2" and (currentExpression[index] != expression[index]):
            return False

    return True


def calculatePx(pattern, currentExpression, possibleSolutions):
    Nmatches = 0
    TotalPossibleExpressions = len(possibleSolutions)

    for expression in possibleSolutions:
        if checkPattern(pattern, currentExpression, expression):
            Nmatches += 1

    Px = Nmatches/TotalPossibleExpressions

    return Px


def calculateInformationVariables(expression, possibleSolutions, patterns):
    """ 
    E[i] = sumatoria(p(x)*log2(1/p(x))) 
    Donde:
    x = patron de similitud
    p(x) = (N de eq posibles con el patron x dado) / (N total de eq) 
    """

    EinfoOfExpression = 0
    for pattern in patterns:
        Px = calculatePx(pattern, expression, possibleSolutions)
        if Px != 0:  # Avoiding division by 0
            EinfoOfExpression += Px*math.log((1/Px), 2)

    return EinfoOfExpression


def main():

    # Parsing the .txt's into a lists

    # voy a tomar como que las unicas valid inputs son las posibles soluciones. Me veo obligado a esto ya que si no achico el set de posibles opciones se apaga el sol antes de que termine de computar todos los valores
    with open("validSolutions8.txt", "r") as f:
        validInputs = []

        for expression in f:
            # Esto porque como la leo de un txt termina con un /n para el salto de linea. lo cual es indeseable
            validInputs.append(expression[:-1])

        pass

    with open("validSolutions8.txt", "r") as f:
        possibleSolutions = []

        for expression in f:
            possibleSolutions.append(expression[:-1])

    # Calculating possible patterns of similitud
    """ 
    Simbolos de patron de similitud
    0: if sign not in eq
    1: if sign in eq but not in that place
    2: if sign in eq in that place
    """
    numberOfTiles = 8  # HARDCODED

    patterns = []

    def foo(l, numberOfTiles):
        yield from itertools.product(*([l] * numberOfTiles))

    for x in foo("012", numberOfTiles):
        patterns.append(x)

    # Calculating information given by every variable
    informationList = []

    total = len(validInputs)  # DEBUGING

    for expression in validInputs:
        print(round(len(informationList)/total, 3) * 100)  # DEBUGING

        EinfoOfExpression = calculateInformationVariables(
            expression, possibleSolutions, patterns)

        print("{} info: {}".format(expression, EinfoOfExpression))

        informationList.append((expression, EinfoOfExpression))

    # Sorting list

    def sortingKey(e):
        return e[1]

    informationList.sort(key=sortingKey)

    createInfoTxt(informationList)

    return


if __name__ == "__main__":
    main()
